ramips: add support for Netgear AC 1200 R6120

Brother-Lal opened the pull request:
Hello!
These commits should add support for the Netgear dual-band router r6120, as well as basic support for the Sercomm/Netgear firmware file format.

It is not 100% in a mergeable state, but i wanted to get feedback on some issues.
(and ofc, all other issues that are still present... :+1: )

Open Questions:
1. The GPIO/LEDs are all working, but the ralink,group is most likely not correct - somebody with more info on the mappings should clarify that.
2. The firmware format is accepted by the nmrp as well as the webinterface, but the webinterface only copies a first part. Anybody has any ideas on that - i am willing to provide flashdumps and other support if needed, as i am out of ideas.

Otherwise I tested all things i know, macs, wifi, ethernet, sysupgrade is all working as it should.

Greetings,

Brother Lal

Brother-Lal committed:
firmware-utils: add sercomm/netgear tool

This adds a tool to generate a firmware file accepted
by Netgear or sercomm devices.

They use a zip-packed rootfs with header and a custom
checksum. The generated Image can be flashed via the
nmrpflash tool or the webinterface of the router.

Signed-off-by: Ludwig Thomeczek <ledesrc@wxorx.net>

Brother-Lal committed:
ramips: add support for Netgear R6120

This patch adds support for the Netgear R6120, aka Netgear AC1200.

Specification:
- SoC: MediaTek MT7628 (580 MHz)
- Flash: 16 MiB
- RAM: 64 MiB
- Wireless: 2.4Ghz(builtin) and 5Ghz (MT7612E)
- LAN speed: 10/100
- LAN ports: 4
- WAN speed: 10/100
- WAN ports: 1
- Serial baud rate of Bootloader and factory firmware: 57600

To flash use nmrpflash with the provided factory.img.
Flashing via webinterface will not work, for now.

Signed-off-by: Ludwig Thomeczek <ledesrc@wxorx.net>

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -298,6 +298,12 @@ mzk-ex750np)
 	set_usb_led "$boardname:blue:status"
 	;;
+netgear,r6120)
+	ucidef_set_led_default "power" "power" "$boardname:green:power" "1"
```
What is the purpose of this line. Let diag.sh handle the led.

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -298,6 +298,12 @@ mzk-ex750np)
 	;;
+netgear,r6120)
+	ucidef_set_led_default "power" "power" "$boardname:green:power" "1"
+	ucidef_set_led_netdev "lan" "lan" "$boardname:green:lan" "eth0.2"
```
depending on the configuration, the lan ports might not be eth0.2. Please use `ucidef_set_led_switch` instead. the last argument is a bitmask of ports to trigger.

But if eth0.2 is your lan port something is wrong. According to your configuration, eth0.1 should be lan.

blogic replied:
@Brother-Lal this needs to be fixed before we can merge it.

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+			gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
```
drop this blank line

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+			gpios = <&gpio1 10 GPIO_ACTIVE_LOW>;
+		};
+
+                wlan_orange {
```
spaces vs. tabs here. please check the whole dts for similar issues.

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+			read-only;
+		};
+
+		nvram: partition@60000 {
```
any reason why this partition isn't marked as read-only as well

Brother-Lal replied:
I though i needed to modify nvram and forgot to add it back, thanks, fixed!

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+			reg = <0xff0000 0x10000>;
+			read-only;
+		};
+
```
drop these and the following blank line

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+			mediatek,mtd-eeprom = <&factory 0x28000>;
+			ieee80211-freq-limit = <5000000 6000000>;
+			mtd-mac-address = <&nvram 0x100B0>;
+			mtd-mac-address-increment = <(2)>;
```
Where is the base mac address incremented by one used?

Brother-Lal replied:
Wan has +1

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+		};
+	};
+};
+
```
blank lines

mkresin commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -90,6 +90,23 @@ define Device/mt7628
+define Device/netgear_r6120
+  DTS := R6120
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := $(ralink_default_fw_size_8M)
```
ralink_default_fw_size_16M

mkresin commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -90,6 +90,23 @@ define Device/mt7628
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := $(ralink_default_fw_size_8M)
+  DEVICE_TITLE := Netgear AC1200 R6120
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport
```
no usb led, no kmod-usb-ledtrig-usbport

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "gpio", "p0led_an", "p1led_an", "p2led_an",
```
drop the i2c and gpio pins from the group. none of the pins is used
![bildschirmfoto vom 2018-06-21 23-48-03](https://user-images.githubusercontent.com/4547948/41747392-2d6ed04e-75ae-11e8-8563-ad32c36c5e04.png)

mkresin commented on target/linux/ramips/dts/R6120.dts:
```diff
@@ -0,0 +1,146 @@
+                        label = "r6120:orange:wlan5g";
+                        gpios = <&gpio1 7 GPIO_ACTIVE_LOW>;
+                };
+
```
drop this blank line

mkresin reviewed (changes requested):
I've only done a quick review. Expect more changes if I've the time do do a full review.

Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

Brother-Lal commented:
Thanks for the review, i'll try to do the request changes soonish.
For the gpio-mappings, is that sheet somewhere available? I had a link at some point to a different png of the mappings, but lost it.
Would be good for the next time to check.

Brother-Lal commented:
Thanks for the docu-sheet!
I fixed the errors, and had a look at the existing dgn35000 firmware file.
Seems its version 0 and the netgear r6120 and r6220 have the version 1, with zipped rootfs.
But i could modify the tool to zip or not zip the rootfs depending on the version number.
I'll try some other things with the firmware generator, to get it to work with the webinterface, and update.
Feel free to point out any other things i can fix in the meantime :)

blogic commented:
patch fails to apply, please rebase it

Brother-Lal commented:
Patch is rebased, and adapted to new power led handling.
Current state: webinterface upload of firmware still fails, and the wifi leds stopped working.
I'll fix the wifi leds soonish, and then i hope we can merge, as flashing via nmrpflash is possible.

I do have the r6220 here also, as soon as i add this support to the firmware tool, i'll try to see and compare if i can flash via webinterface or what is wrong with the generated file. (In a separate PR)

Brother-Lal commented:
Updated and fixed wifi leds, switch declaration.
Should be mergeable now. @mkresin 

Webinterface flash is still not working, but will be worked on later.

mkresin commented:
I fix the remaining issues of the PR and pushed the cleanup to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary).

Either I broke something while doing the cleanup or there is a bug in the image code since the beginning. If I build an image, the following files are added to my local tree:

- target/linux/ramips/image/fs.zip
- target/linux/ramips/image/rootfs.img

What ever the reason for this behaviour is, it is wrong and need to be fixed.

Brother-Lal commented:
Yes, this is a big, thise are temporary files that should be deleted after successful

Brother-Lal commented:
Wrong button on mobile, sry...
Those are temp files that i forgot to delete in the image generation script.
I'll fix it this afternoon.

mkresin commented:
Beside of that, shouldn't the files stored in a temporary directory? Possibly in the same directory like the input file?

Brother-Lal commented:
Yes, they should be stored in a temp directory - I thought the image generation happens in such a directory, as the tempfile is just generated inside the current working dir.
I probably would then need to extract the current image generation path from the path for the sysupgrade image and prepend that...
Or would a `cd $(TMPDIR)` before the mksercommfw tool execution in the image makefile be an ok solution, if the files get erased after that?

mkresin commented:
> Or would a cd $(TMPDIR) before the mksercommfw tool execution in the image makefile be an ok solution, if the files get erased after that?

No. Please have a look how the other fw-utils are doing it.

Brother-Lal commented:
updated to remove the files and create them in the tmp directory in the first place.
also makes the mv @ in the image makefile unnecessary.

mkresin commented:
Would have been easier for both of us if you had updated the PR with the code from my staging tree before applying more changes.

I updated the code in my [staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) with the changes you have done. Hopefully I didn't missed anything.

Please build and runtime test an image build from [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) to ensure I imported all changes and everything still works as it should be.

Brother-Lal commented:
Works as intended! Thank you for persevering on that one ...

mkresin commented:
Glad to hear. Added the commits to my push queue and will commit them to master later the day.

adrianschmutzler commented:
@Brother-Lal Just stumbled about R6120 not having MAC+1 used. From earlier comments I found that this would be the WAN MAC. Is there a reason why this one is not set in 02_network? Otherwise, I would send a fix for that.

Brother-Lal commented:
@adrianschmutzler Sadly i neither got my notes nor the hardware still here, so i can't really comment.
There were some additionall changes in between and somebody else integrated and fixed further on that hw.
But it can very realistic be an oversight on my part that i forgot to integrate that correctly.
Would be good to confirm that with original HW and fix if necessary. I will see if i can figure out who has access to one,

adrianschmutzler commented:
@Brother-Lal Save your time. I've realized that there is a default case to set wan_mac to eth0 + 1. So, your device has had correct WAN MAC address from the beginning. Sorry for the confusion. (Although I dislike the idea of having a default case for stuff like that.)

rotanid commented:
@blocktrron has the device :)
